File: mobile/ios/AlMatchLifeCore/Sources/AlMatchLifeCore/AuthenticatedAPIClient.swift
import Foundation

public struct SessionTokens: Codable, Sendable, Equatable {
    public let accessToken: String
    public let refreshToken: String

    public init(accessToken: String, refreshToken: String) {
        self.accessToken = accessToken
        self.refreshToken = refreshToken
    }

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
    }
}

public enum APIClientError: Error, Sendable, Equatable {
    case insecureBaseURL
    case unauthenticated
    case invalidResponse
    case httpStatus(Int)
    case invalidSyntheticSpeech
}

public protocol HTTPTransport: Sendable {
    func data(for request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

public struct URLSessionHTTPTransport: HTTPTransport, Sendable {
    private let session: URLSession

    public init(session: URLSession = .shared) { self.session = session }

    public func data(for request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIClientError.invalidResponse }
        return (data, http)
    }
}

public struct SyntheticSpeech: Sendable {
    public let data: Data
    public let captionID: String
    public let language: String
    public let voice: String
}

public struct CallSignalAcknowledgement: Decodable, Sendable {
    public let ok: Bool
    public let eventID: String
    public let duplicate: Bool
    enum CodingKeys: String, CodingKey { case ok, duplicate; case eventID = "event_id" }
}

public actor AuthenticatedAPIClient {
    private static let tokenAccount = "mobile-session-v1"
    private let baseURL: URL
    private let transport: any HTTPTransport
    private let tokenStore: any SecureTokenStoring
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private var refreshTask: Task<SessionTokens, Error>?

    public init(
        baseURL: URL,
        transport: any HTTPTransport = URLSessionHTTPTransport(),
        tokenStore: any SecureTokenStoring = KeychainTokenStore()
    ) throws {
        let isLocalhost = ["localhost", "127.0.0.1", "::1"].contains(baseURL.host?.lowercased() ?? "")
        guard baseURL.scheme?.lowercased() == "https" || (isLocalhost && baseURL.scheme == "http") else {
            throw APIClientError.insecureBaseURL
        }
        self.baseURL = baseURL
        self.transport = transport
        self.tokenStore = tokenStore
    }

    public func install(tokens: SessionTokens) throws {
        let data = try encoder.encode(tokens)
        guard let encoded = String(data: data, encoding: .utf8) else { throw APIClientError.invalidResponse }
        try tokenStore.save(encoded, account: Self.tokenAccount)
    }

    public func clearSession() throws {
        try tokenStore.delete(account: Self.tokenAccount)
    }

    public func bootstrap() async throws -> MobileBootstrap {
        let (data, _) = try await authenticatedRequest(path: "/api/mobile/bootstrap", method: "GET")
        return try decoder.decode(MobileBootstrap.self, from: data)
    }

    public func createRealtimeSession(callID: String, otherEmail: String, callType: String) async throws -> RealtimeSession {
        let body = try JSONSerialization.data(withJSONObject: ["other_email": otherEmail, "call_type": callType])
        let path = "/api/calls/\(encodePath(callID))/translation/realtime-session"
        let (data, _) = try await authenticatedRequest(path: path, method: "POST", body: body)
        return try decoder.decode(RealtimeSessionEnvelope.self, from: data).session
    }

    public func translatedSpeech(
        callID: String, captionID: String, otherEmail: String, callType: String, voice: String = "coral"
    ) async throws -> SyntheticSpeech {
        let body = try JSONSerialization.data(withJSONObject: [
            "other_email": otherEmail, "call_type": callType, "voice": voice,
        ])
        let path = "/api/calls/\(encodePath(callID))/captions/\(encodePath(captionID))/speech"
        let (data, response) = try await authenticatedRequest(path: path, method: "POST", body: body)
        guard response.value(forHTTPHeaderField: "X-AI-Generated-Voice") == "true",
              response.mimeType == "audio/mpeg",
              let responseCaptionID = response.value(forHTTPHeaderField: "X-Caption-Id"),
              responseCaptionID == captionID,
              let language = response.value(forHTTPHeaderField: "Content-Language"),
              let responseVoice = response.value(forHTTPHeaderField: "X-AI-Voice"),
              responseVoice == voice,
              !language.isEmpty,
              !data.isEmpty else {
            throw APIClientError.invalidSyntheticSpeech
        }
        return SyntheticSpeech(data: data, captionID: responseCaptionID, language: language, voice: responseVoice)
    }

    public func sendCallSignal(
        callID: String, otherEmail: String, callType: NativeCallType,
        type: String, eventID: String, payload: [String: String]
    ) async throws -> CallSignalAcknowledgement {
        let body = try JSONSerialization.data(withJSONObject: [
            "other_email": otherEmail, "call_type": callType.rawValue,
            "type": type, "event_id": eventID, "payload": payload,
        ])
        let path = "/api/calls/\(encodePath(callID))/signals"
        let (data, _) = try await authenticatedRequest(path: path, method: "POST", body: body)
        let acknowledgement = try decoder.decode(CallSignalAcknowledgement.self, from: data)
        guard acknowledgement.ok, acknowledgement.eventID == eventID else { throw APIClientError.invalidResponse }
        return acknowledgement
    }

    public func iceConfiguration(callID: String, otherEmail: String, callType: NativeCallType) async throws -> NativeICEConfiguration {
        let path = callPath(callID, suffix: "ice-servers", otherEmail: otherEmail, callType: callType, after: nil)
        let (data, _) = try await authenticatedRequest(path: path, method: "GET")
        return try decoder.decode(NativeICEConfiguration.self, from: data)
    }

    public func pollCallSignals(
        callID: String, otherEmail: String, callType: NativeCallType, after: Double
    ) async throws -> NativeSignalPoll {
        let path = callPath(callID, suffix: "signals", otherEmail: otherEmail, callType: callType, after: after)
        let (data, _) = try await authenticatedRequest(path: path, method: "GET")
        return try decoder.decode(NativeSignalPoll.self, from: data)
    }

    public func acknowledgeCallSignals(
        callID: String, otherEmail: String, callType: NativeCallType, eventIDs: [String]
    ) async throws -> NativeSignalACK {
        let body = try JSONSerialization.data(withJSONObject: [
            "other_email": otherEmail, "call_type": callType.rawValue, "event_ids": Array(eventIDs.prefix(50)),
        ])
        let path = "/api/calls/\(encodePath(callID))/signals/ack"
        let (data, _) = try await authenticatedRequest(path: path, method: "POST", body: body)
        return try decoder.decode(NativeSignalACK.self, from: data)
    }

    public func sendSessionDescription(
        callID: String, otherEmail: String, callType: NativeCallType, type: String, sdp: String, eventID: String
    ) async throws -> CallSignalAcknowledgement {
        let body = try JSONSerialization.data(withJSONObject: [
            "other_email": otherEmail, "call_type": callType.rawValue, "type": type, "event_id": eventID,
            "payload": ["type": type, "sdp": sdp, "call_type": callType.rawValue],
        ])
        let (data, _) = try await authenticatedRequest(
            path: "/api/calls/\(encodePath(callID))/signals", method: "POST", body: body
        )
        return try decoder.decode(CallSignalAcknowledgement.self, from: data)
    }

    public func sendICECandidate(
        callID: String, otherEmail: String, callType: NativeCallType,
        candidate: NativeICECandidate, eventID: String
    ) async throws -> CallSignalAcknowledgement {
        var candidatePayload: [String: Any] = ["candidate": candidate.candidate, "call_type": callType.rawValue]
        if let value = candidate.sdpMid { candidatePayload["sdpMid"] = value }
        if let value = candidate.sdpMLineIndex { candidatePayload["sdpMLineIndex"] = value }
        if let value = candidate.usernameFragment { candidatePayload["usernameFragment"] = value }
        let body = try JSONSerialization.data(withJSONObject: [
            "other_email": otherEmail, "call_type": callType.rawValue, "type": "ice",
            "event_id": eventID, "payload": candidatePayload,
        ])
        let (data, _) = try await authenticatedRequest(
            path: "/api/calls/\(encodePath(callID))/signals", method: "POST", body: body
        )
        return try decoder.decode(CallSignalAcknowledgement.self, from: data)
    }

    private func authenticatedRequest(
        path: String, method: String, body: Data? = nil, mayRefresh: Bool = true
    ) async throws -> (Data, HTTPURLResponse) {
        let tokens = try loadTokens()
        var request = try makeRequest(path: path, method: method, body: body)
        request.setValue("Bearer \(tokens.accessToken)", forHTTPHeaderField: "Authorization")
        let result = try await transport.data(for: request)
        if result.1.statusCode == 401 && mayRefresh {
            let refreshed = try await refreshTokens(staleAccessToken: tokens.accessToken)
            var retry = try makeRequest(path: path, method: method, body: body)
            retry.setValue("Bearer \(refreshed.accessToken)", forHTTPHeaderField: "Authorization")
            let retryResult = try await transport.data(for: retry)
            guard (200..<300).contains(retryResult.1.statusCode) else {
                throw APIClientError.httpStatus(retryResult.1.statusCode)
            }
            return retryResult
        }
        guard (200..<300).contains(result.1.statusCode) else { throw APIClientError.httpStatus(result.1.statusCode) }
        return result
    }

    private func refreshTokens(staleAccessToken: String) async throws -> SessionTokens {
        let current = try loadTokens()
        if current.accessToken != staleAccessToken { return current }
        if let refreshTask { return try await refreshTask.value }
        let baseURL = self.baseURL
        let transport = self.transport
        let refreshToken = current.refreshToken
        let task = Task<SessionTokens, Error> {
            var request = URLRequest(url: baseURL.appending(path: "/api/auth/refresh"))
            request.httpMethod = "POST"
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            request.httpBody = try JSONSerialization.data(withJSONObject: ["refresh_token": refreshToken])
            let (data, response) = try await transport.data(for: request)
            guard (200..<300).contains(response.statusCode) else { throw APIClientError.unauthenticated }
            return try JSONDecoder().decode(SessionTokens.self, from: data)
        }
        refreshTask = task
        do {
            let tokens = try await task.value
            try install(tokens: tokens)
            refreshTask = nil
            return tokens
        } catch {
            refreshTask = nil
            try? clearSession()
            throw error
        }
    }

    private func loadTokens() throws -> SessionTokens {
        guard let encoded = try tokenStore.read(account: Self.tokenAccount),
              let data = encoded.data(using: .utf8) else { throw APIClientError.unauthenticated }
        return try decoder.decode(SessionTokens.self, from: data)
    }

    private func makeRequest(path: String, method: String, body: Data?) throws -> URLRequest {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL,
              url.host == baseURL.host else { throw APIClientError.invalidResponse }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")
        if let body {
            request.httpBody = body
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        return request
    }

    private func callPath(
        _ callID: String, suffix: String, otherEmail: String, callType: NativeCallType, after: Double?
    ) -> String {
        var components = URLComponents()
        components.path = "/api/calls/\(encodePath(callID))/\(suffix)"
        var items = [URLQueryItem(name: "other_email", value: otherEmail), URLQueryItem(name: "call_type", value: callType.rawValue)]
        if let after { items.append(URLQueryItem(name: "after", value: String(after))) }
        components.queryItems = items
        return components.string ?? ""
    }

    private func encodePath(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? ""
    }
}

File: mobile/android/AlMatchLifeCore/src/main/kotlin/com/almatchlife/core/AuthenticatedApiClient.kt
package com.almatchlife.core

import java.net.URI
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

internal fun <T> failedFutureCompat(failure: Throwable): CompletableFuture<T> =
    CompletableFuture<T>().also { it.completeExceptionally(failure) }

data class SessionTokens(val accessToken: String, val refreshToken: String)

interface SessionTokenStore {
    fun save(tokens: SessionTokens)
    fun read(): SessionTokens?
    fun clear()
}

data class ApiRequest(
    val method: String,
    val uri: URI,
    val headers: Map<String, String>,
    val body: ByteArray? = null,
)

data class ApiResponse(
    val statusCode: Int,
    val headers: Map<String, String>,
    val body: ByteArray,
)

interface ApiTransport {
    fun execute(request: ApiRequest): CompletableFuture<ApiResponse>
}

data class SignalAcknowledgement(val ok: Boolean, val eventId: String)

data class RealtimeSession(
    val clientSecret: String,
    val expiresAt: Long,
    val model: String,
    val transcriptionModel: String,
    val transport: String,
    val callsEndpoint: URI,
    val sourceLanguage: String,
)

data class CaptionSegment(
    val id: String,
    val text: String,
    val speakerEmail: String,
    val sourceLanguage: String,
    val createdAt: Double,
    val sequence: Long = 0,
)

data class CaptionPoll(val captions: List<CaptionSegment>, val serverTime: Double)
data class CaptionTranslation(val captionId: String, val translatedText: String, val targetLanguage: String)
data class SyntheticSpeech(
    val bytes: ByteArray,
    val captionId: String,
    val language: String,
    val voice: String,
)

interface MobileWireCodec {
    fun encodeRefreshRequest(refreshToken: String): ByteArray
    fun decodeSessionTokens(body: ByteArray): SessionTokens
    fun encodeCallSignal(
        callId: String,
        otherEmail: String,
        callType: NativeCallType,
        type: String,
        eventId: String,
        reason: String?,
    ): ByteArray
    fun decodeSignalAcknowledgement(body: ByteArray): SignalAcknowledgement
    fun decodeIceConfiguration(body: ByteArray): IceConfiguration
    fun decodeSignalPoll(body: ByteArray): SignalPoll
    fun encodeSessionDescription(payload: VoipCallPayload, type: String, sdp: String, eventId: String): ByteArray
    fun encodeIceCandidate(payload: VoipCallPayload, candidate: IceCandidate, eventId: String): ByteArray
    fun encodeSignalAcknowledgement(payload: VoipCallPayload, eventIds: List<String>): ByteArray
    fun decodeAcknowledgedEventIds(body: ByteArray): List<String>
    fun encodeCallContext(otherEmail: String, callType: NativeCallType, voice: String? = null): ByteArray
    fun encodeCaptionTranslationRequest(otherEmail: String, callType: NativeCallType, targetLanguage: String): ByteArray
    fun encodeCaption(segment: CaptionSegment, otherEmail: String, callType: NativeCallType): ByteArray
    fun decodePublishedCaption(body: ByteArray): CaptionSegment
    fun decodeCaptionPoll(body: ByteArray): CaptionPoll
    fun decodeCaptionTranslation(body: ByteArray): CaptionTranslation
    fun decodeRealtimeSession(body: ByteArray): RealtimeSession
}

class ApiClientException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)

class AuthenticatedApiClient(
    baseUri: URI,
    private val transport: ApiTransport,
    private val tokenStore: SessionTokenStore,
    private val codec: MobileWireCodec,
    allowLoopbackHttp: Boolean = false,
) : CallSignaling {
    private val baseUri = validateBaseUri(baseUri, allowLoopbackHttp)
    private val refreshLock = Any()
    private var refreshInFlight: CompletableFuture<SessionTokens>? = null

    fun request(
        path: String,
        method: String = "GET",
        body: ByteArray? = null,
        headers: Map<String, String> = emptyMap(),
        query: Map<String, String> = emptyMap(),
    ): CompletableFuture<ApiResponse> {
        val tokens = tokenStore.read()
            ?: return failedFutureCompat(ApiClientException("authentication required"))
        return authorizedRequest(path, method, body, headers, query, tokens.accessToken).thenCompose { response ->
            if (response.statusCode != 401) CompletableFuture.completedFuture(response)
            else refreshSingleFlight(tokens.accessToken).thenCompose { refreshed ->
                authorizedRequest(path, method, body, headers, query, refreshed.accessToken)
            }
        }
    }

    override suspend fun send(
        type: String,
        eventId: String,
        payload: VoipCallPayload,
        reason: String?,
    ) {
        requireIdentifier(eventId, 16, 80, "event id")
        val body = codec.encodeCallSignal(
            payload.callId,
            payload.callerEmail,
            payload.callType,
            type,
            eventId,
            reason,
        )
        val response = request(
            path = "/api/calls/${payload.callId}/signals",
            method = "POST",
            body = body,
            headers = mapOf("Content-Type" to "application/json"),
        ).await()
        if (response.statusCode !in 200..299) throw ApiClientException("signal rejected: ${response.statusCode}")
        val acknowledgement = codec.decodeSignalAcknowledgement(response.body)
        if (!acknowledgement.ok || acknowledgement.eventId != eventId) {
            throw ApiClientException("signal acknowledgement mismatch")
        }
    }

    suspend fun createRealtimeSession(payload: VoipCallPayload): RealtimeSession {
        val response = checkedRequest(
            "/api/calls/${safePath(payload.callId)}/translation/realtime-session",
            codec.encodeCallContext(payload.callerEmail, payload.callType),
        )
        return codec.decodeRealtimeSession(response.body)
    }

    suspend fun publishCaption(payload: VoipCallPayload, segment: CaptionSegment): CaptionSegment {
        if (segment.text.isBlank() || segment.text.length > 500 || segment.sequence < 0) {
            throw ApiClientException("invalid caption segment")
        }
        val response = checkedRequest(
            "/api/calls/${safePath(payload.callId)}/captions",
            codec.encodeCaption(segment, payload.callerEmail, payload.callType),
        )
        return codec.decodePublishedCaption(response.body)
    }

    suspend fun pollCaptions(payload: VoipCallPayload, after: Double): CaptionPoll {
        val response = request(
            path = "/api/calls/${safePath(payload.callId)}/captions",
            query = mapOf(
                "other_email" to payload.callerEmail,
                "call_type" to payload.callType.wireValue,
                "after" to after.coerceAtLeast(0.0).toString(),
            ),
        ).await()
        requireSuccess(response, "caption poll")
        return codec.decodeCaptionPoll(response.body)
    }

    suspend fun translateCaption(
        payload: VoipCallPayload,
        captionId: String,
        targetLanguage: String,
    ): CaptionTranslation {
        requireIdentifier(captionId, 8, 80, "caption id")
        val body = codec.encodeCaptionTranslationRequest(payload.callerEmail, payload.callType, targetLanguage)
        val response = checkedRequest(
            "/api/calls/${safePath(payload.callId)}/captions/${safePath(captionId)}/translation",
            body,
        )
        return codec.decodeCaptionTranslation(response.body).copy(captionId = captionId)
    }

    suspend fun translatedSpeech(
        payload: VoipCallPayload,
        captionId: String,
        voice: String = "coral",
    ): SyntheticSpeech {
        requireIdentifier(captionId, 8, 80, "caption id")
        if (voice !in ALLOWED_SYNTHETIC_VOICES) throw ApiClientException("unsupported synthetic voice")
        val response = checkedRequest(
            "/api/calls/${safePath(payload.callId)}/captions/${safePath(captionId)}/speech",
            codec.encodeCallContext(payload.callerEmail, payload.callType, voice),
            accept = "audio/mpeg",
        )
        val contentType = response.header("Content-Type")?.substringBefore(';')?.trim()?.lowercase()
        val generated = response.header("X-AI-Generated-Voice")
        val responseCaptionId = response.header("X-Caption-Id")
        val language = response.header("Content-Language")?.trim().orEmpty()
        val responseVoice = response.header("X-AI-Voice")
        if (contentType != "audio/mpeg" || generated != "true" || responseCaptionId != captionId ||
            responseVoice != voice || language.isBlank() || response.body.isEmpty() ||
            response.body.size > MAX_SYNTHETIC_SPEECH_BYTES
        ) throw ApiClientException("invalid synthetic speech response")
        return SyntheticSpeech(response.body, captionId, language, voice)
    }

    private suspend fun checkedRequest(path: String, body: ByteArray, accept: String = "application/json"): ApiResponse {
        val response = request(
            path = path,
            method = "POST",
            body = body,
            headers = mapOf("Content-Type" to "application/json", "Accept" to accept),
        ).await()
        requireSuccess(response, "request")
        return response
    }

    private fun requireSuccess(response: ApiResponse, operation: String) {
        if (response.statusCode !in 200..299) throw ApiClientException("$operation rejected: ${response.statusCode}")
    }

    private fun ApiResponse.header(name: String): String? =
        headers.entries.firstOrNull { it.key.equals(name, ignoreCase = true) }?.value

    private fun safePath(value: String): String {
        requireIdentifier(value, 8, 128, "path identifier")
        return value
    }

    private fun authorizedRequest(
        path: String,
        method: String,
        body: ByteArray?,
        headers: Map<String, String>,
        query: Map<String, String>,
        accessToken: String,
    ): CompletableFuture<ApiResponse> {
        requireSafeToken(accessToken)
        val normalizedMethod = method.uppercase()
        if (normalizedMethod !in setOf("GET", "POST", "PUT", "PATCH", "DELETE")) {
            return failedFutureCompat(ApiClientException("unsupported HTTP method"))
        }
        val uri = endpoint(path, query)
        val safeHeaders = headers.toMutableMap()
        if (safeHeaders.keys.any { it.equals("Authorization", ignoreCase = true) }) {
            return failedFutureCompat(ApiClientException("authorization header is managed internally"))
        }
        safeHeaders["Authorization"] = "Bearer $accessToken"
        if (safeHeaders.keys.none { it.equals("Accept", ignoreCase = true) }) {
            safeHeaders["Accept"] = "application/json"
        }
        return transport.execute(ApiRequest(normalizedMethod, uri, safeHeaders, body))
    }

    private fun refreshSingleFlight(staleAccessToken: String): CompletableFuture<SessionTokens> = synchronized(refreshLock) {
        refreshInFlight?.let { return@synchronized it }
        val current = tokenStore.read()
            ?: return@synchronized failedFutureCompat(ApiClientException("authentication required"))
        if (current.accessToken != staleAccessToken) return@synchronized CompletableFuture.completedFuture(current)
        requireSafeToken(current.refreshToken)
        val request = ApiRequest(
            method = "POST",
            uri = endpoint("/api/auth/refresh"),
            headers = mapOf("Accept" to "application/json", "Content-Type" to "application/json"),
            body = codec.encodeRefreshRequest(current.refreshToken),
        )
        val candidate = transport.execute(request).thenApply { response ->
            if (response.statusCode !in 200..299) throw ApiClientException("refresh rejected")
            codec.decodeSessionTokens(response.body).also {
                requireSafeToken(it.accessToken)
                requireSafeToken(it.refreshToken)
                tokenStore.save(it)
            }
        }
        refreshInFlight = candidate
        candidate.whenComplete { _, failure ->
            synchronized(refreshLock) {
                if (refreshInFlight === candidate) {
                    refreshInFlight = null
                    if (failure != null) tokenStore.clear()
                }
            }
        }
        candidate
    }

    private fun endpoint(path: String, query: Map<String, String> = emptyMap()): URI {
        if (!path.startsWith('/') || path.startsWith("//") || '?' in path || '#' in path) {
            throw ApiClientException("invalid API path")
        }
        if (query.size > 16 || query.any { (key, value) -> key.isBlank() || key.length > 64 || value.length > 2048 }) {
            throw ApiClientException("invalid API query")
        }
        val queryText = query.entries.joinToString("&") { (key, value) ->
            "${encodeQueryValue(key)}=${encodeQueryValue(value)}"
        }.ifEmpty { null }
        val originPath = URI(baseUri.scheme, null, baseUri.host, baseUri.port, path, null, null).toASCIIString()
        val resolved = URI.create(if (queryText == null) originPath else "$originPath?$queryText")
        if (resolved.scheme != baseUri.scheme || resolved.host != baseUri.host || resolved.port != baseUri.port) {
            throw ApiClientException("API endpoint escaped origin")
        }
        return resolved
    }

    private fun encodeQueryValue(value: String): String = java.net.URLEncoder
        .encode(value, Charsets.UTF_8.name())
        .replace("+", "%20")

    private fun requireSafeToken(token: String) {
        if (token.isBlank() || token.length > 8192 || '\r' in token || '\n' in token) {
            throw ApiClientException("invalid bearer token")
        }
    }

    private fun requireIdentifier(value: String, minimum: Int, maximum: Int, label: String) {
        if (value.length !in minimum..maximum || !value.all { it.isLetterOrDigit() || it == '_' || it == '-' }) {
            throw ApiClientException("invalid $label")
        }
    }

    private suspend fun <T> CompletableFuture<T>.await(): T = suspendCoroutine { continuation ->
        whenComplete { value, failure ->
            if (failure == null) continuation.resume(value)
            else continuation.resumeWithException(
                if (failure is CompletionException && failure.cause != null) failure.cause!! else failure,
            )
        }
    }

    private companion object {
        const val MAX_SYNTHETIC_SPEECH_BYTES = 2 * 1024 * 1024
        val ALLOWED_SYNTHETIC_VOICES = setOf("alloy", "ash", "ballad", "coral", "echo", "sage", "shimmer", "verse")
        fun validateBaseUri(uri: URI, allowLoopbackHttp: Boolean): URI {
            // 10.0.2.2 is the Android Emulator's reserved alias for the host loopback.
            // It is accepted only when the caller explicitly enables debug loopback HTTP.
            val loopback = uri.host?.lowercase() in setOf("localhost", "127.0.0.1", "::1", "10.0.2.2")
            if (uri.scheme != "https" && !(allowLoopbackHttp && uri.scheme == "http" && loopback)) {
                throw ApiClientException("HTTPS is required")
            }
            if (uri.host.isNullOrBlank() || uri.userInfo != null || uri.query != null || uri.fragment != null) {
                throw ApiClientException("invalid base URI")
            }
            return URI(uri.scheme, null, uri.host, uri.port, "/", null, null)
        }
    }
}
